lib/_http_agent.js

```diff
@@ -392,13 +419,31 @@ Agent.prototype.removeSocket = function removeSocket(s, options) {
         // Don't leak
         if (sockets[name].length === 0)
           delete sockets[name];
+        this.totalSocketCount--;
       }
     }
   }
 
+  let req;
   if (this.requests[name] && this.requests[name].length) {
     debug('removeSocket, have a request, make a socket');
-    const req = this.requests[name][0];
+    req = this.requests[name][0];
+  } else {
+    // TODO(rickyes): this logic will not be FIFO across origins.
+    // There might be older requests in a different origin, but
+    // if the origin which releases the socket has pending requests
+    // that will be prioritized.
+    for (const prop in this.requests) {
+      debug('removeSocket, have a request with different origin,' +
+        ' make a socket');
+      req = this.requests[prop][0];
+      options = req[kRequestOptions];
```

Comment:
I'm not sure this will work? It bypass the per origin logic, e.g. what if this specific origin is already at maxSockets?

Reply:
There is indeed this problem, maybe we can add a check whether this `specific origin` is already at `maxSockets` ?

Reply:
I think we need some tests that hit this issue, and fix it afterwards.

Reply:
Okay, I'll add some tests to cover this issue tomorrow morning.

Reply:
I added the check.